Optimizer/FcTypeChecker.hs

```diff
@@ -219,19 +212,336 @@ ensureIdenticalTypes types = unless (go types) $ throwError "Type mismatch in ca
+  (ty, pgm') <- setCtxM ctx $ tcFcOptProgram pgm
+  return (ty, FcPgmValDecl bind' pgm')
+-- ^ type check program term, wrap into FcPgmTerm
+tcFcOptProgram (FcPgmTerm tm) = second FcPgmTerm <$> tcFcOptTerm tm
+
+tcFcOptTerm :: FcOptTerm -> FcM (FcType, FcResTerm)
+-- ^ Type check a type abstraction
+tcFcOptTerm (FcOptTmTyAbs as t) = do
+  tyVarsNotInFcCtxM as
+  (ty, t') <- extendCtxTysM as (map kindOf as) (tcFcOptTerm t)
+  return (fcTyAbs as ty, t')
+-- ^ Type check a type application (fallback, TODO)
+tcFcOptTerm (FcOptTmTyApp tm tys) = do
+  bind <- mkFcResBind tm
+  res_ty <- tcFcOptTyApp (fval_bind_ty bind) tys
+  let tyapp_r = FcResTmApp (FcRatorVar (fval_bind_var bind)) (map FcAtType tys)
+  return (res_ty, FcResTmLet [bind] tyapp_r)
+-- ^ Type check a term application
+tcFcOptTerm (FcOptTmApp t ts) = tcFcOptTmApp t ts
+-- ^ Type check a case statement
+tcFcOptTerm (FcOptTmCase tm alts) = do
+  (ty, tm_r) <- tcFcOptTerm tm
+  (tys, alts_r) <- case alts of
+    (FcAAlts alts') -> second FcAAlts . unzip <$> mapM (tcFcOptAAlt ty) alts'
+    (FcPAlts alts') -> second FcPAlts . unzip <$> mapM (tcFcOptPAlt ty) alts'
+  ensureIdenticalTypes tys
+  return (head tys, FcResTmCase tm_r alts_r)
+-- ^ Type check a local let binding
+tcFcOptTerm (FcOptTmLet bind t) = do
+  (ctx, bind_r) <- tcFcOptBind bind
+  second (FcResTmLet [bind_r]) <$> setCtxM ctx (tcFcOptTerm t)
+-- ^ Type check an abstraction (fallback)
+tcFcOptTerm (FcOptTmAbs vs tm) = do
+  bind <- mkFcResBind (FcOptTmAbs vs tm)
+  return (fval_bind_ty bind, FcResTmLet [bind] (FcResTmApp (FcRatorVar $ fval_bind_var bind) []))
+tcFcOptTerm (FcOptTmVar x) = lookupTmVarM x >>= \ty -> return (ty, FcResTmApp (FcRatorVar x) [])
+-- ^ Type check a primitive literal
+tcFcOptTerm (FcOptTmPrim (PrimLitTm lit)) = tcPrimLit lit >>= \ty -> return (ty, FcResTmLit lit)
+-- ^ Type check a datacon (fallback for dc arity 0)
+tcFcOptTerm (FcOptTmDataCon dc) = do
+  (as, arg_tys, dc_tc) <- lookupDataConTyM dc
+  unless (null arg_tys) $
+    throwErrorM (text "tcFcOptTmApp" <+> colon <+> text "Unsaturated datacon application")
+  return (mkDataConTy (as, arg_tys, dc_tc), FcResTmApp (FcRatorCon dc) [])
+
+
+-- | Type check an optimizer value binding.
+tcFcOptBind :: FcOptBind -> FcM (FcCtx, FcResBind)
+tcFcOptBind (FcBind x ty tm) = do
+  tmVarNotInFcCtxM x  -- GEORGE: Ensure is not already bound
+  kind <- tcType ty
+  unless (kind == KStar) $
+    throwError "tcFcValBind: Kind mismatch (FcValBind)"
+  (ty', ab) <- case tm of     -- Type check bound term and put it into FcResAbs
+    (FcOptTmAbs vs tm') -> do
+      let (xs, tys) = unzip vs
+      (ty_tm, tm_r) <- extendCtxTmsM (x:xs) (ty:tys) (tcFcOptTerm tm')
+      let ty_ab = foldr mkFcArrowTy ty_tm tys
+      return (ty_ab, FcResAbs vs tm_r)
+    tm'                 -> do
+      (ty_tm, tm_r) <- extendCtxTmM x ty (tcFcOptTerm tm')
+      return (ty_tm, FcResAbs [] tm_r)
+  unless (ty `eqFcTypes` ty') $ throwErrorM (text "Global let type doesnt match:"
```

Comment:
I guess this links back to our discussing earlier today, but do you really want to force these types to be syntactically identical?